### PR TITLE
Add ability to use 'free_space' plugin on remote server

### PR DESCRIPTION
**Pull request by metaMMA**
### Motivation for changes:
Some users run flexget locally, but have remote storage. This amendment allows the user to take advantage of the `free_space` features when using remote storage.
### Detailed changes:
- Create ssh connection to remote server
- Poll for free space using either `df` or `du` (depending on whether disk space is shared)

### Config usage if relevant (new plugin or updated schema):
```
tasks:
  test_free_space:
    free_space:
      host: 'subdomain.hostingcompany.com'
      user: 'Cool_username'
      path: '/media/sdr1/Cool_username/'
      space: 60000
      ssh_key_filepath: '/home/local_username/.ssh/id_rsa'
      allotment: 1000000
    rss: https://distrowatch.com/news/torrents.xml
    accept_all: yes
    regexp:
      reject_excluding:
        -  "sparkylinux*"
    make_rss:
      file: /home/local_username/Documents/test_output.rss
```
### Config wiki info:
```
      host: <remote server URL or IP address>
      user: <remote server login username>
      path: <remote server path. Can be any existing path (including '/') if entire disk is available for use. Must be your 'home' directory if sharing disk space>
      space: <minimum space (in MB) that must be available to accept new entries>
      ssh_key_filepath: <local machine public key filepath. See link below>
      allotment: <OPTIONAL: Allocated disk space (in MB) (only needed if sharing disk space)>
      port: <OPTIONAL: SSH port for remote server. Defaults to '22'>
```
[SSH key info](https://www.tecmint.com/ssh-passwordless-login-using-ssh-keygen-in-5-easy-steps/)
### Log and/or tests output (preferably both):
```
(flexget) username@hostname:~/flexget$ flexget execute
2020-02-23 02:56:53 VERBOSE  task_queue                    There are 1 tasks to execute. Shutdown will commence when they have completed.
2020-02-23 02:56:54 VERBOSE  details       test_free_space Produced 25 entries.
2020-02-23 02:56:54 VERBOSE  task          test_free_space REJECTED: `pclinuxos64-kde5-2020.02.iso.torrent` by regexp plugin because regexp 'sparkylinux*' didn't match
2020-02-23 02:56:54 VERBOSE  task          test_free_space REJECTED: `untangle_1500_x64.iso.torrent` by regexp plugin because regexp 'sparkylinux*' didn't match
2020-02-23 02:56:54 VERBOSE  task          test_free_space REJECTED: `neon-user-20200220-1117.iso.torrent` by regexp plugin because regexp 'sparkylinux*' didn't match
2020-02-23 02:56:54 VERBOSE  task          test_free_space REJECTED: `NetBSD-9.0-amd64.iso.torrent` by regexp plugin because regexp 'sparkylinux*' didn't match
2020-02-23 02:56:54 VERBOSE  task          test_free_space REJECTED: `raspbian-2020-02-13-raspbian-buster.zip.torrent` by regexp plugin because regexp 'sparkylinux*' didn't match
2020-02-23 02:56:54 VERBOSE  task          test_free_space REJECTED: `robolinux64-mate3d-v10.9.iso.torrent` by regexp plugin because regexp 'sparkylinux*' didn't match
2020-02-23 02:56:54 VERBOSE  task          test_free_space REJECTED: `neon-user-20200213-1120.iso.torrent` by regexp plugin because regexp 'sparkylinux*' didn't match
2020-02-23 02:56:54 VERBOSE  task          test_free_space REJECTED: `xubuntu-18.04.4-desktop-amd64.iso.torrent` by regexp plugin because regexp 'sparkylinux*' didn't match
2020-02-23 02:56:54 VERBOSE  task          test_free_space REJECTED: `tails-amd64-4.3.iso.torrent` by regexp plugin because regexp 'sparkylinux*' didn't match
2020-02-23 02:56:54 VERBOSE  task          test_free_space REJECTED: `Archman_KDE_20200209.iso.torrent` by regexp plugin because regexp 'sparkylinux*' didn't match
2020-02-23 02:56:54 VERBOSE  task          test_free_space REJECTED: `4MLinux-31.2-64bit.iso.torrent` by regexp plugin because regexp 'sparkylinux*' didn't match
2020-02-23 02:56:54 VERBOSE  task          test_free_space REJECTED: `debian-10.3.0-amd64-xfce-CD-1.iso.torrent` by regexp plugin because regexp 'sparkylinux*' didn't match
2020-02-23 02:56:54 VERBOSE  task          test_free_space ACCEPTED: `sparkylinux-2020.02-x86_64-rescue.iso.torrent` by accept_all plugin
2020-02-23 02:56:54 VERBOSE  task          test_free_space ACCEPTED: `sparkylinux-2020.02-x86_64-multimedia.iso.torrent` by accept_all plugin
2020-02-23 02:56:54 VERBOSE  task          test_free_space ACCEPTED: `sparkylinux-2020.02-x86_64-gameover.iso.torrent` by accept_all plugin
2020-02-23 02:56:54 VERBOSE  task          test_free_space ACCEPTED: `sparkylinux-2020.02-i686-rescue.iso.torrent` by accept_all plugin
2020-02-23 02:56:54 VERBOSE  task          test_free_space ACCEPTED: `sparkylinux-2020.02.1-x86_64-xfce.iso.torrent` by accept_all plugin
2020-02-23 02:56:54 VERBOSE  task          test_free_space ACCEPTED: `sparkylinux-2020.02.1-x86_64-minimalgui.iso.torrent` by accept_all plugin
2020-02-23 02:56:54 VERBOSE  task          test_free_space ACCEPTED: `sparkylinux-2020.02.1-x86_64-mate.iso.torrent` by accept_all plugin
2020-02-23 02:56:54 VERBOSE  task          test_free_space ACCEPTED: `sparkylinux-2020.02.1-x86_64-lxqt.iso.torrent` by accept_all plugin
2020-02-23 02:56:54 VERBOSE  task          test_free_space ACCEPTED: `sparkylinux-2020.02-x86_64-xfce.iso.torrent` by accept_all plugin
2020-02-23 02:56:54 VERBOSE  task          test_free_space ACCEPTED: `sparkylinux-2020.02-x86_64-minimalgui.iso.torrent` by accept_all plugin
2020-02-23 02:56:54 VERBOSE  task          test_free_space ACCEPTED: `sparkylinux-2020.02-x86_64-minimalcli.iso.torrent` by accept_all plugin
2020-02-23 02:56:54 VERBOSE  task          test_free_space ACCEPTED: `sparkylinux-2020.02-x86_64-mate.iso.torrent` by accept_all plugin
2020-02-23 02:56:54 VERBOSE  task          test_free_space ACCEPTED: `sparkylinux-2020.02-x86_64-lxqt.iso.torrent` by accept_all plugin
2020-02-23 02:56:54 INFO     paramiko.transport                 Connected (version 2.0, client OpenSSH_7.4p1)
2020-02-23 02:56:56 INFO     paramiko.transport                 Authentication (publickey) successful!
2020-02-23 02:56:57 ERROR    free_space    test_free_space Less than 60000 MB of free space in /media/sdr1/Cool_username/ aborting task.
Exception ignored in: <function BufferedFile.__del__ at 0x7f7e7b5e1d08>
Traceback (most recent call last):
  File "/home/local_username/flexget/lib/python3.7/site-packages/paramiko/file.py", line 66, in __del__
  File "/home/local_username/flexget/lib/python3.7/site-packages/paramiko/channel.py", line 1392, in close
  File "/home/local_username/flexget/lib/python3.7/site-packages/paramiko/channel.py", line 991, in shutdown_write
  File "/home/local_username/flexget/lib/python3.7/site-packages/paramiko/channel.py", line 963, in shutdown
  File "/home/local_username/flexget/lib/python3.7/site-packages/paramiko/channel.py", line 1246, in _send_eof
  File "/home/local_username/flexget/lib/python3.7/site-packages/paramiko/message.py", line 232, in add_int
TypeError: 'NoneType' object is not callable

```
#### To Do:

- [ ] Find way to suppress paramiko exception ([reference](https://github.com/paramiko/paramiko/issues/1078))
- [ ] Update flexget.com wiki

### CODE REVIEW

**Comment by paranoidi** [comment]
This should probably be a separate plugin.

**Comment by metaMMA** [comment]
> This should probably be a separate plugin.

I could do that, but I'd just be copying and pasting the entire plugin and adding the 24 lines that allow me to check a disk on a remote server. Unless, you are envisioning something else that I'm not able to imagine.

**Comment by paranoidi** [comment]
then at least redundant remote: true should be removed?

**Comment by metaMMA** [comment]
> then at least redundant remote: true should be removed?

Yes. It is redundant. Removed :)

**Comment by paranoidi** [comment]
I do not understand the meaning or usage of "allotment" option.

**Comment by metaMMA** [comment]
> I do not understand the meaning or usage of "allotment" option.

Yes, I meant to add some info to the 'wiki' area. Many seedboxes do 'disk sharing'. For example, `/media/sda1` may be 6TB in size and have 5 or 6 users, each with their own home directory and 1TB allotment (1000000 MB): 
```/media/sda1/user1
/media/sda1/user2
/media/sda1/user3
/media/sda1/user4
/media/sda1/user5```
When this is true, we can't simply check the entire disk for free space. So, they would use the path to their home directory: `/media/sda1/user3` and use `allotment: 1000000`. The code will then find the disk usage of their home directory and calculate free space by subtracting the `disk usage` from their `allotment`.

Does that make sense?

**Comment by metaMMA** [comment]
@gazpachoking @liiight @paranoidi Anything else I need to clear up on this one?

**Comment by paranoidi** [comment]
Looks good to me

File: flexget/plugins/operate/free_space.py [hunk 1]
```diff
@@ -8,18 +8,42 @@
...
+        resp = ''.join(outlines)
+        ssh.close()
+        try:
+            if config['allotment'] != -1:
+                free = int(config['allotment']) - ((int(resp.strip()) * 1024) / 1000000)
+            else:
+                free = (int(resp.strip()) / 1000)
+        except ValueError:
+            logger.error('Non-integer was returned when calcualting disk usage.')
+            return "ERR"
```
[the remark this answers is not in the document]
Comment (metaMMA) [reply flexget/plugins/operate/free_space.py]: That makes sense. Removed the `return "ERR"` from the two locations and simply abort task at that point.

**Comment by gazpachoking** [comment]
Looks good, thanks!

**Pull request merged**